src/clients/node/src/node.zig

```diff
@@ -15,783 +19,472 @@ const CreateTransfersResult = tb.CreateTransfersResult;
+                    );
+
+                    if (std.mem.eql(u8, field.name, "timestamp") and value != 0) {
+                        return translate.throw(
+                            env,
+                            "Timestamp should be set to 0 as this will be set correctly by Replica.",
+                        );
+                    }
+
+                    @field(event, field.name) = switch (@typeInfo(field.type)) {
+                        .Struct => @as(field.type, @bitCast(value)),
+                        else => value,
+                    };
+                }
+            },
+            u128 => event.* = try translate.u128_from_value(env, object, "lookup"),
+            else => @compileError("invalid Event type"),
+        }
     }
 
-    if (argc != 1) translate.throw(
-        env,
-        "Function tick() requires 1 argument exactly.",
-    ) catch return null;
+    // Return only the memory for the Events. The total size of the bytes allocation can be inferred
+    // via `bytes[0..max(bytes.len, events.len * ResultFrom(Event))]`.
+    return std.mem.sliceAsBytes(events);
+}
 
-    const context_raw = translate.value_external(
-        env,
-        argv[0],
-        "Failed to get Client Context pointer.",
-    ) catch return null;
-    const context = contextCast(context_raw.?) catch return null;
+// For a given op, packet memory is allocated to hold
+// `max(sizeof(Event/Request) * n, sizeof(Result/Response))`. As long as the op and n are known, the
+// allocation size can be computed.
+//
+// When submitting the packet, the data_size is `sizeof(Event) * n`. On packet results, we overwrite
+// the memory with `sizeof(Result) * completed` where completed <= n. If less Results are completed
+// than Events, the completed size no longer reflects "n". So instead, data_size stores
+// `packed(n, completed)`: arg1 to know alloc size, arg2 to know what to parse back to JS.
+const DataSize = packed struct(u32) {
+    event_count: u16,
+    result_count: u16,
+};
 
-    context.client.tick();
-    context.io.tick() catch |err| switch (err) {
-        // TODO exhaustive switch
-        else => {
-            translate.throw(env, "Failed to tick IO.") catch return null;
-        },
+fn encode_results_into_data(
+    comptime Event: type,
+    comptime Result: type,
+    result_bytes: []const u8,
+    event_data: []align(16) u8,
+) DataSize {
+    // Event data should have been allocated with enough space to also contain the Results.
+    const event_count = @divExact(event_data.len, @sizeOf(Event));
+    const results = std.mem.bytesAsSlice(Result, event_data.ptr[0 .. @sizeOf(Result) * event_count]);
+
+    const packet_results = std.mem.bytesAsSlice(Result, result_bytes);
+    assert(packet_results.len <= results.len);
+    @memcpy(results[0..packet_results.len], packet_results);
+
+    return .{
+        .event_count = @intCast(event_count),
+        .result_count = @intCast(packet_results.len),
```

Comment:
Suggestions:

- Make names consistent `results_bytes` and `events_bytes` instead of `events_data`.
- Don't introduce new names like `packet_results`, instead let's just use `buffer` and naturally call it `results`.
- Explain why we're doing it in the comments.

```suggestion
fn encode_results_into_data(
    comptime Event: type,
    comptime Result: type,
    result_bytes: []const u8,
    event_bytes: []align(16) u8,
) BufferSize {
    const event_count = @divExact(event_bytes.len, @sizeOf(Event));
    const buffer: []Result = buffer: {
        // The results must be stored until the JS callback is called, we use the same buffer for
        // both events and results.
        // Reinterpreting `event_bytes` is safe since the buffer should have been allocated with
        // enough space to also contain the Results.
        const buffer_size = @sizeOf(Result) * event_count;
        break :buffer std.mem.bytesAsSlice(Result, event_bytes.ptr[0..buffer_size]);
    };

    const results = std.mem.bytesAsSlice(Result, result_bytes);
    assert(results.len <= buffer.len);
    @memcpy(buffer[0..results.len], results);

    return .{
        .event_count = @intCast(event_count),
        .result_count = @intCast(results.len),
```